### DISCUSSION TITLE
How to enable mTLS on S3 service?

### DISCUSSION THREAD

**Discussion by jameskhedley** [seaweedfs/seaweedfs#4901] [Q&A] [answered]
Hi, question about securing access to the S3 server using mutual TLS.

I can see the following options for s3:

```
  -s3.cert.file string
    	path to the TLS certificate file
  -s3.key.file string
    	path to the TLS private key file
```
So I thought I could use those to enable mTLS. The S3 server starts fine like this: 

```
$ ./weed server -dir=/tmp/seaweed_data -s3 -s3.cert.file=./certs/public.crt -s3.key.file=./certs/private.key
```

But then when I curl the S3 endpoint it lets me in without the key or cert:

```
$ curl -k https://localhost:8333
<?xml version="1.0" encoding="UTF-8"?>
<ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><DisplayName></DisplayName><ID></ID></Owner><Buckets><Bucket><CreationDate>2023-10-03T14:47:11Z</CreationDate><Name>backups</Name></Bucket></Buckets></ListAllMyBucketsResult>
```

Basically I want to deploy this in an environment where we use mTLS with cert-manager to do everything. We could use credentials instead of course but that means some extra plumbing.

**Comment by kmlebedev** [comment]
Hi, what version are you using?
Can you find a similar line in the logs?
```
I1009 16:14:33.068281 s3.go:294 Start Seaweed S3 API Server 30GB 3.57  at https port 8000
```
Are you sure you didn't use the optional parameter `-s3.port.https`?
 
```
weed -v 4 server -filer -s3 -s3.key.file=./docker/compose/tls/volume01.dev.key -s3.cert.file=./docker/compose/tls/volume01.dev.crt
```
Locally TLS is enabled, perhaps you were missing the `-filer` parameter.

**Reply by jameskhedley** [reply]
Thanks for the quick reply!

> Hi, what version are you using? 

```
 ((3.57)) ~/github/seaweedfs/weed$ go run weed.go version
version 30GB 3.57  linux amd64
```

> perhaps you were missing the -filer parameter

Tried it but made no difference.

Actually I lied, I was running by `go run` from a checked out tag, would that make a difference?

Actual full command I just ran was:

`$ go run weed.go server -dir=/tmp/seaweed_data -filer -s3 -s3.cert.file=./certs/public.crt -s3.key.file=./certs/private.key`

> I1009 12:35:54.997741 s3.go:294 Start Seaweed S3 API Server 30GB 3.57  at https port 8333

So I think https is there, as you can see in the curl, it just doesn't seem to do anything with the cert and key - afaik those are only needed if the server is verifying the client (aka mTLS), otherwise it's just the client verifying the server's cert so the key isn't necessary.

Curl with -v shows more TLS information:

```
$ curl -kv https://localhost:8333
*   Trying 127.0.0.1:8333...
* Connected to localhost (127.0.0.1) port 8333 (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: CN=backup-oc24.local-minio
*  start date: May  2 19:26:55 2023 GMT
*  expire date: May  1 19:26:55 2026 GMT
*  issuer: CN=Elastic Certificate Tool Autogenerated CA
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* using HTTP/1.1
> GET / HTTP/1.1
> Host: localhost:8333
> User-Agent: curl/8.0.1
> Accept: */*
> 
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
< HTTP/1.1 200 OK
< Accept-Ranges: bytes
< Content-Length: 292
< Content-Type: application/xml
< Server: SeaweedFS S3
< X-Amz-Request-Id: 1696851551925423829
< Date: Mon, 09 Oct 2023 11:39:11 GMT
```
If I try with http it fails of course.

**Comment by kmlebedev** [comment]
It looks like you want to use client certificates for s3 server side authorization?

If yes, then I don’t understand how you plan to configure (pass the client certificate) a native s3 client, for example boto3. According to the documentation there is no such option.
https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#using-environment-variables

**Comment by jameskhedley** [comment]
Yes, maybe terminology is confusing.

I want to have to specify cert and key on both the client and server side,
otherwise the request is rejected. So then I don't really need credentials.

I can't understand why the options exist to have a server side cert and key
if they're never needed. If the client doesn't need to specify then the
server only needs cert bundle I think.

**Comment by kmlebedev** [comment, marked as answer]
> Yes, maybe terminology is confusing.


Then this is a custom modification of the s3 API interface, since its standard does not provide for authorization using a client certificate; instead, AWS Signature Version 4 is used. https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html

Accordingly, you can send your PR or use an external proxy as an ingress, which can do mTLS, for example caddy https://gist.github.com/mojzu/b093d79e73e7aa302dde8e335945b2cd

**Reply by jameskhedley** [reply]
Thanks, I will try to figure out a PR, if not I can try Caddy or maybe OpenResty.

**Reply by jameskhedley** [reply]
https://github.com/seaweedfs/seaweedfs/pull/4910

It probably needs changes, lmk what you think!
